README.md
# IP To CIDR (ip2cidr)

by Ron Dilley <ron.dilley@uberadmin.com>

You can find the latest information on ip2cidr [here](http://www.uberadmin.com/Projects/ip2cidr/ "IP To CIDR")

## What is IP To CIDR (ip2cidr)?

IP To CIDR is a small and fast utility for consolidating lists of IP addresses to
CIDR blocks.  It is pretty fast and was built to work with 10M+ IP lists.  If you
have a large IP address list and you want to summarize to network ranges based on
how densely populated the CIDR is, this tool may be able to help.

## Why use it?

I built this tool to solve a specific problem with large lists of known bad IP
addresses that needed to be loaded into an ACL for blocking.  Unfortunately, the
firewall I was using only supported a limited number of CIDRs.  To make it work,
I needed to consolidate IP addresses into CIDR blocks based on the number of
IP addresses in a given CIDR.

## Implementation

IP To CIDR has a simple command lines interface.  In it's
simplest form, pass a text file as an argument and the output
will be all of the lines that are not recognized, any consolidated
CIDRs and whatever IPs are left over (converted to /32 notation).

```
ip2cidr v0.5 [Jul 21 2023 - 21:50:24]

syntax: ip2cidr [options] filename [filename ...]
 -d|--debug (0-9)       enable debugging info
 -h|--help              this info
 -H|--hbit {bits}       max network bits (default: 31)
 -l|--lbit {bits}       min network bits (default: 24)
 -t|--thold {percent}   consolidation threshold (default: 51)
 -v|--version           display version information
 -V|--verbose           show additional information
 filename               one or more files to process, use '-' to read from stdin
```

The debug option is most useful when the tool is compiled
with the --ENABLE-DEBUG switch.

ip2cidr processes and consolidates IPv4 addresses in octet and CIDR format.  IPv6
and malformed IP addresses are not consolidated but are passed on to STDOUT without
any changes.  CIDR blocks that are larger than the minimum mask (see -l|--lbit) are
not consolidated but are passed on to STDOUT without any changes.  If a CIDR block
does not start with the first address (e.g., the host id is non-zero), ip2cidr will
not consolidate and instead will pass the CIDR block to STDOUT without any changes.
IPv4 CIDR blocks with /32 will be converted to a host IP address in octet format
and will be consolidated.

Use the verbose switch (see -V|--verbose) to get more information about which entries
in a file are consolidated and which are passed to STDOUT without any change.

A typical run of ip2cidr is to pass the IP file as an
argument.

```
% ./ip2cidr ip_list.txt > consolidated_ip_list.txt
```

Or you can pass the IP addresses via STDIN.

```
% cat ip_list.txt | ./ip2cidr > consolidated_ip_list.txt
```

You control the minimum and maximum bitmask along with the percentage of IPs
in a particular CIDR using command line arguments as indicated in the --help.

To consolidate CIDRs starting at /20 and ending at /30 when at least 75% of
the IP addresses exist in a given CIDR, use the following arguments:

```
% ./ip2cidr -l 20 -H 30 -t 75 ip_list.txt > consolidated_ip_list_20_to_30_at_75.txt
```

## Security Implications

Assume that there are errors in the ip2cidr source that
would allow a specially crafted logs to allow an attacker
to exploit this to gain access to the computer that it is
running on!!!  Don't trust this software and install and use
it at your own risk.

## Bugs

I am not a programmer by any stretch of the imagination.  I
have attempted to remove the obvious bugs and other
programmer related errors but please keep in mind the first
sentence.  If you find an issue with code, please send me
an e-mail with details and I will be happy to look into
it.

Ron Dilley
ron.dilley@uberadmin.com


NEWS
Version 0.1 (May 08, 2023):

* First public release.

Version 0.2 (May 17, 2023):

* All core features for IPv4 complete.

Version 0.3 (May 20, 2023):

* Minor changes to verbose output
* Porting to MinGW

Version 0.4 (July 21, 2023):

* Added CIDR block input processing and consolidation

Version 0.5 (July 21, 2023):

* Minor bug fix on invalid CIDR output
* Added comments on why items were not consolidated



src/main.c
/*****
 *
 * Description: Main Functions
 *
 * BSD 3-Clause License
 *
 * Copyright (c) 2008-2023, Ron Dilley
 * All rights reserved.
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 ****/

/****
 *
 * includes
 *.
 ****/

#include "main.h"

/****
 *
 * local variables
 *
 ****/

/****
 *
 * global variables
 *
 ****/

PUBLIC int quit = FALSE;
PUBLIC int reload = FALSE;
PUBLIC Config_t *config = NULL;

/****
 *
 * external variables
 *
 ****/

extern int errno;
extern char **environ;

/****
 *
 * main function
 *
 ****/

int main(int argc, char *argv[])
{
  PRIVATE int c = 0;

#ifndef DEBUG
# ifndef MINGW
  struct rlimit rlim;

  rlim.rlim_cur = rlim.rlim_max = 0;
  setrlimit(RLIMIT_CORE, &rlim);
# endif
#endif

  /* setup config */
  config = (Config_t *)XMALLOC(sizeof(Config_t));
  XMEMSET(config, 0, sizeof(Config_t));

  /* force mode to forground */
  config->mode = MODE_INTERACTIVE;

  /* store current pid */
  config->cur_pid = getpid();

  while (1)
  {
    int this_option_optind = optind ? optind : 1;
#ifdef HAVE_GETOPT_LONG
    int option_index = 0;
    static struct option long_options[] = {
        {"verbose", no_argument, 0, 'V'},
                {"version", no_argument, 0, 'v'},
        {"debug", required_argument, 0, 'd'},
        {"help", no_argument, 0, 'h'},
        {"hbit", required_argument, 0, 'H'},
        {"lbit", required_argument, 0, 'l'},
        {"thold", required_argument, 0, 't'},
        {0, no_argument, 0, 0}};
    c = getopt_long(argc, argv, "Vvd:hH:l:t:", long_options, &option_index);
#else
    c = getopt(argc, argv, "Vvd:hH:l:t:");
#endif

    if (c EQ - 1)
      break;

    switch (c)
    {

    case 'v':
      /* show the version */
      print_version();
      return (EXIT_SUCCESS);

    case 'V':
      /* enable verbose mode */
      config->verbose = TRUE;
      break;
      
    case 'd':
      /* show debig info */
      config->debug = atoi(optarg);
      break;

    case 'h':
      /* show help info */
      print_help();
      return (EXIT_SUCCESS);

    case 'H':
      /* max network bits */
      config->maxBits = atoi(optarg);
      break;

    case 'l':
      /* min network bits */
      config->minBits = atoi(optarg);
      break;

    case 't':
      /* consolidation threshold */
      config->threshold = atof(optarg) / 100;
      break;

    default:
      fprintf(stderr, "Unknown option code [0%o]\n", c);
    }
  }

  /* set required defaults if not defined */
  if (config->threshold EQ 0)
    config->threshold = DEFAULT_THRESHOLD;

  if ( config->minBits EQ 0)
    config->minBits = DEFAULT_MIN_BITS;

  if (config->maxBits EQ 0)
    config->maxBits = DEFAULT_MAX_BITS;

  /*
   * get to work
   */

  /* process all the files */
  while (optind < argc)
  {
    processFile(argv[optind++]);
  }

  /*
   * finished with the work
   */

  cleanup();

  return (EXIT_SUCCESS);
}

/****
 *
 * display prog info
 *
 ****/

void show_info(void)
{
  fprintf(stderr, "%s v%s [%s - %s]\n", PROGNAME, VERSION, __DATE__, __TIME__);
  fprintf(stderr, "By: Ron Dilley\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "%s comes with ABSOLUTELY NO WARRANTY.\n", PROGNAME);
  fprintf(stderr, "This is free software, and you are welcome\n");
  fprintf(stderr, "to redistribute it under certain conditions;\n");
  fprintf(stderr, "See the GNU General Public License for details.\n");
  fprintf(stderr, "\n");
}

/*****
 *
 * display version info
 *
 *****/

PRIVATE void print_version(void)
{
  printf("%s v%s [%s - %s]\n", PROGNAME, VERSION, __DATE__, __TIME__);
}

/*****
 *
 * print help info
 *
 *****/

PRIVATE void print_help(void)
{
  print_version();

  fprintf(stderr, "\n");
  fprintf(stderr, "syntax: %s [options] filename [filename ...]\n", PACKAGE);

#ifdef HAVE_GETOPT_LONG
  fprintf(stderr, " -d|--debug (0-9)       enable debugging info\n");
  fprintf(stderr, " -h|--help              this info\n");
  fprintf(stderr, " -H|--hbit {bits}       max network bits (default: 31)\n");
  fprintf(stderr, " -l|--lbit {bits}       min network bits (default: 24)\n");
  fprintf(stderr, " -t|--thold {percent}   consolidation threshold (default: 51)\n");
  fprintf(stderr, " -v|--version           display version information\n");
  fprintf(stderr, " -V|--verbose           show additional information\n");
  fprintf(stderr, " filename               one or more files to process, use '-' to read from stdin\n");
#else
  fprintf(stderr, " -d {lvl}       enable debugging info\n");
  fprintf(stderr, " -h             this info\n");
  fprintf(stderr, " -H {bits}      max network bits (default: 31)\n");
  fprintf(stderr, " -l {bits}      min network bits (default: 24)\n");
  fprintf(stderr, " -t {percent}   consolidation threshold (default: 51)\n");
  fprintf(stderr, " -v             display version information\n");
  fprintf(stderr, " -V             show additional information\n");
  fprintf(stderr, " filename       one or more files to process, use '-' to read from stdin\n");
#endif

  fprintf(stderr, "\n");
}

/****
 *
 * cleanup
 *
 ****/

PRIVATE void cleanup(void)
{
#ifdef MEM_DEBUG
  XFREE_ALL();
#else
  XFREE(config);
#endif
}
